openhtf/util/console_output.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2018 Google Inc. All Rights Reserved.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Console output utilities for OpenHTF.
+
+This module provides a few conviences to help formatting output for the CLI, as
+well as for forking output to both a logger (for use with test record loggers)
+and to the CLI directly (through sys.stdout).
+"""
+
+import logging
+import math
+import sys
+import time
+
+import colorama
+import contextlib2 as contextlib
+
+# Colorama module has to be initialized before use.
+colorama.init()
+
+_LOG = logging.getLogger(__name__)
+
+
+class ActionFailedError(Exception):
+  """Raised (and caught) when an action is marked as failed."""
+
+
+def banner_print(msg, color='', width=60, file=sys.stdout):
+  """Print the message as a banner with a fixed width.fixed
+
+  Args:
+    msg: The message to print.
+    color: Optional colorama color string to be applied to the message. You can
+        concatenate colorama color strings together in order to get any set of
+        effects you want.
+    width: Total width for the resulting banner.
+    file: A file object to which the banner text will be written. Intended for
+        use with CLI output file objects like sys.stdout.
+
+  Example:
+
+    >>> banner_print('Foo Bar Baz')
+    ======================== Foo Bar Baz =======================
+  """
+  lpad = int(math.ceil((width - len(msg) - 2) / 2.0)) * '='
+  rpad = int(math.floor((width - len(msg) - 2) / 2.0)) * '='
+  file.write('{color}{lpad} {msg} {rpad}\n'.format(
+      color=color, lpad=lpad, msg=msg, rpad=rpad))
+  file.write(colorama.Style.RESET_ALL)
```

Comment:
probably doesn't matter, but I think it would look a bit better if this was {reset} in the format string above and was written before the \n.  if you had a particular reason for doing it after the \n then it's fine as-is.

Reply:
Done.